Readme.md
# nlcst-link-modifier [![Build Status](https://img.shields.io/travis/wooorm/nlcst-link-modifier.svg?style=flat)](https://travis-ci.org/wooorm/nlcst-link-modifier) [![Coverage Status](https://img.shields.io/coveralls/wooorm/nlcst-link-modifier.svg?style=flat)](https://coveralls.io/r/wooorm/nlcst-link-modifier?branch=master)

Classify links as `LinkNode`s.

Implemented by [retext-link](https://github.com/wooorm/retext-link), but separated for use by standalone (non-[retext](https://github.com/wooorm/retext)) parsers.

> Note: this project is useful in combination with natural language parsers like [parse-latin](https://github.com/wooorm/parse-latin), [parse-dutch](https://github.com/wooorm/parse-dutch), and [parse-english](https://github.com/wooorm/parse-english).

## Installation

npm:
```sh
$ npm install nlcst-link-modifier
```

Component.js:
```sh
$ component install wooorm/nlcst-link-modifier
```

Bower:
```sh
$ bower install nlcst-link-modifier
```

## Usage

```js
var modifier = require('nlcst-link-modifier');
var ParseEnglish = require('parse-english');
var english = new ParseEnglish();

/* Attach the modifier. */
modifier(english);

english.parse(
    'Who doesn’t like http://example.com? You? google.com.'
).children[0].children;
```

Yields:

```json
[
  {
    "type": "SentenceNode",
    "children": [
      {
        "type": "WordNode",
        "children": [
          {
            "type": "TextNode",
            "value": "Who"
          }
        ]
      },
      {
        "type": "WhiteSpaceNode",
        "value": " "
      },
      {
        "type": "WordNode",
        "children": [
          {
            "type": "TextNode",
            "value": "doesn"
          },
          {
            "type": "PunctuationNode",
            "value": "’"
          },
          {
            "type": "TextNode",
            "value": "t"
          }
        ]
      },
      {
        "type": "WhiteSpaceNode",
        "value": " "
      },
      {
        "type": "WordNode",
        "children": [
          {
            "type": "TextNode",
            "value": "like"
          }
        ]
      },
      {
        "type": "WhiteSpaceNode",
        "value": " "
      },
      {
        "type": "LinkNode",
        "value": "http://example.com"
      },
      {
        "type": "PunctuationNode",
        "value": "?"
      }
    ]
  },
  {
    "type": "WhiteSpaceNode",
    "value": " "
  },
  {
    "type": "SentenceNode",
    "children": [
      {
        "type": "WordNode",
        "children": [
          {
            "type": "TextNode",
            "value": "You"
          }
        ]
      },
      {
        "type": "PunctuationNode",
        "value": "?"
      },
      {
        "type": "WhiteSpaceNode",
        "value": " "
      },
      {
        "type": "LinkNode",
        "value": "google.com"
      },
      {
        "type": "PunctuationNode",
        "value": "."
      }
    ]
  }
]
```

## Performance

On a MacBook Air, **parse-english** performs about 8% slower on content filled with links, and a 7% slower on content without links, when using this modifier.

```
             parse w/ modifier
  1,554 op/s » A paragraph (5 sentences, 100 words, 10 links)
  2,453 op/s » A paragraph (5 sentences, 100 words, no links)

             parse w/o modifier
  1,702 op/s » A paragraph (5 sentences, 100 words, 10 links)
  2,640 op/s » A paragraph (5 sentences, 100 words, no links)
```

## Related

- [nlcst](https://github.com/wooorm/nlcst)
- [nlcst-emoji-modifier](https://github.com/wooorm/nlcst-emoji-modifier)
- [parse-latin](https://github.com/wooorm/parse-latin)
- [parse-dutch](https://github.com/wooorm/parse-dutch)
- [parse-english](https://github.com/wooorm/parse-english)
- [retext](https://github.com/wooorm/retext)
- [textom](https://github.com/wooorm/textom)

## License

MIT © [Titus Wormer](http://wooorm.com)


History.md

0.1.1 / 2014-12-14
==================

 * Update benchmark results in `Readme.md`
 * Add link to personal website to copyright in `Readme.md`
 * Refactor npm script targets in `package.json`
 * Update eslint, parse-english
 * Fix malformed `bower.json`

0.1.0 / 2014-11-19
==================

 * Update for changes in parse-latin


test/index.js
'use strict';

/**
 * Dependencies.
 */

var linkModifier,
    ParseEnglish,
    correctURLs,
    pseudoCorrectURLs,
    incorrectURLs,
    assert;

linkModifier = require('../');
ParseEnglish = require('parse-english');
correctURLs = require('./fixtures/correct');
pseudoCorrectURLs = require('./fixtures/pseudo-correct');
incorrectURLs = require('./fixtures/incorrect');
assert = require('assert');

/**
 * `ParseEnglish`.
 */

var parseEnglish;

parseEnglish = new ParseEnglish();

linkModifier(parseEnglish);

/**
 * Fixtures: Add `pseudoCorrectURLs` to `correctURLs`.
 */

correctURLs = correctURLs.concat(pseudoCorrectURLs);

/**
 * Tests.
 */

describe('nlcst-link-modifier()', function () {
    it('should be a `function`', function () {
        assert(typeof linkModifier === 'function');
    });

    it('should throw when not given a parser', function () {
        assert.throws(function () {
            linkModifier({});
        }, /not a valid parser/);

        assert.doesNotThrow(function () {
            linkModifier(new ParseEnglish());
        });
    });

    it('should classify links such as `www.example.com`', function () {
        var tree,
            node;

        tree = parseEnglish.parse('Check out www.example.com!');

        node = tree.children[0].children[0].children[4];

        assert(node.type === 'LinkNode');
        assert(node.value === 'www.example.com');
    });
});

/**
 * Validate a URL.
 *
 * @param {string} valueBefore
 * @param {string} valueAfter
 * @param {number} index - the location of the URL.
 * @param {string} url
 */

function validateURL(valueBefore, valueAfter, index, url) {
    it('should classify `' + url + '` as a link', function () {
        var tree,
            node;

        tree = parseEnglish.parse(valueBefore + url + valueAfter);

        node = tree.children[0].children[0].children[index];

        assert(node.type === 'LinkNode');
        assert(node.value === url);
    });
}

/**
 * Validate a value is NOT classified as a URL.
 * Logs helpful messages.
 *
 * @param {string} valueBefore
 * @param {string} valueAfter
 * @param {number} index - the location of the non-URL.
 * @param {string} url
 */

function validateIncorrectURL(valueBefore, valueAfter, index, url) {
    it('should NOT classify `' + url + '` as a link', function () {
        var tree,
            paragraph;

        tree = parseEnglish.parse(valueBefore + url + valueAfter);

        paragraph = tree.children[0];

        /* istanbul ignore next */
        paragraph.children.forEach(function (sentence) {
            sentence.children.forEach(function (node) {
                if (node.type === 'LinkNode') {
                    console.log(
                        '  - A URL was, however, ' +
                        'detected: ' + node.value
                    );
                }
            });
        });

        assert(paragraph.children[0].children[index].type !== 'LinkNode');
    });
}

describe('Correct urls', function () {
    correctURLs.forEach(function (correctURL) {
        validateURL('Check out ', ' it\'s awesome!', 4, correctURL);
    });
});

describe('Incorrect urls', function () {
    incorrectURLs.forEach(function (incorrectURL) {
        validateIncorrectURL('Check out ', ' it\'s invalid', 4, incorrectURL);
    });
});

describe('Correct urls suffixed by a full-stop', function () {
    correctURLs.forEach(function (correctURL) {
        validateURL('Check out ', '.', 4, correctURL);
    });
});

describe('Incorrect urls suffixed by a full-stop', function () {
    incorrectURLs.forEach(function (incorrectURL) {
        validateIncorrectURL('Check the invalid URL ', '.', 8, incorrectURL);
    });
});

describe('Correct urls suffixed by a comma', function () {
    correctURLs.forEach(function (correctURL) {
        validateURL('Check out ', ', or that.', 4, correctURL);
    });
});

describe('Incorrect urls suffixed by a comma', function () {
    incorrectURLs.forEach(function (incorrectURL) {
        validateIncorrectURL('Check out ', ', its invalid.', 4, incorrectURL);
    });
});

describe('Correct urls suffixed by multiple full-stops', function () {
    correctURLs.forEach(function (correctURL) {
        validateURL('Check out ', '...', 4, correctURL);
    });
});

describe('Incorrect urls suffixed by multiple full-stops', function () {
    incorrectURLs.forEach(function (incorrectURL) {
        validateIncorrectURL('Check out ', '...', 4, incorrectURL);
    });
});


test/fixtures/correct.js
var correct;

correct = [
    'http://foo.com/blah_blah',
    'http://foo.com/blah_blah/',
    'http://foo.com/blah_blah_(wikipedia)',
    'http://foo.com/blah_blah_(wikipedia)_(again)',
    'http://www.example.com/wpstyle/?p=364',
    'https://www.example.com/foo/?bar=baz&inga=42&quux',
    'http://✪df.ws/123',
    'http://userid:password@example.com:8080',
    'http://userid:password@example.com:8080/',
    'http://userid@example.com',
    'http://userid@example.com/',
    'http://userid@example.com:8080',
    'http://userid@example.com:8080/',
    'http://userid:password@example.com',
    'http://userid:password@example.com/',
    'http://142.42.1.1/',
    'http://142.42.1.1:8080/',
    'http://➡.ws/䨹',
    'http://⌘.ws',
    'http://⌘.ws/',
    'http://foo.com/blah_(wikipedia)#cite-1',
    'http://foo.com/blah_(wikipedia)_blah#cite-1',
    'http://foo.com/unicode_(✪)_in_parens',
    'http://foo.com/(something)?after=parens',
    'http://☺.damowmow.com/',
    'http://code.google.com/events/#&product=browser',
    'http://j.mp',
    'ftp://foo.bar/baz',
    'http://foo.bar/?q=Test%20URL-encoded%20stuff',
    'http://مثال.إختبار',
    'http://例子.测试',
    'http://उदाहरण.परीक्षा',
    'http://-.~_!$&\'()*+,;=:%40:80%2f::::::@example.com',
    'http://1337.net',
    'http://a.b-c.de',
    'http://223.255.255.254'
];

module.exports = correct;


test/fixtures/incorrect.js
var incorrect;

incorrect = [
  '://',
  'http://',
  'http://.',
  'http://..',
  'http://../',
  'http://?',
  'http://??',
  'http://??/',
  'http://#',
  'http://##',
  'http://##/',
  '//',
  '//a',
  '///a',
  '///',
  'http:///a',
  'rdar://1234',
  'h://test',
  'http://3628126748'
];

/* Although technically incorrect, I dont really care about these edge
 * cases:
 *
 * 'ftps://foo.bar/',
 * 'http://-error-.invalid/',
 * 'http://a.b--c.de/',
 * 'http://-a.b.co',
 * 'http://a.b-.co',
 * 'http://0.0.0.0',
 * 'http://10.1.1.0',
 * 'http://10.1.1.255',
 * 'http://224.1.1.1',
 * 'http://1.1.1.1.1',
 * 'http://123.123.123',
 * 'http://.www.foo.bar/',
 * 'http://www.foo.bar./',
 * 'http://.www.foo.bar./',
 * 'http://10.1.1.1',
 * 'http://10.1.1.254'
 */

module.exports = incorrect;


index.js
'use strict';

/**
 * Dependencies.
 */

var nlcstToString;

nlcstToString = require('nlcst-to-string');

/**
 * Constants: node types.
 */

var LINK_NODE;

LINK_NODE = 'LinkNode';

/**
 * Constants.
 */

var EXPRESSION_SLASH;

EXPRESSION_SLASH = /^\/{1,3}$/;

/**
 * Merge punctuation marks and words into a source node.
 *
 * @param {CSTNode} child
 * @param {number} index
 * @param {CSTNode} parent
 * @return {undefined|number} - Either void, or the next index to iterate
 *   over.
 */

function mergeLinks(child, index, parent) {
    var siblings,
        nodes,
        start,
        end,
        currentIndex,
        value,
        type;

    siblings = parent.children;
    nodes = [child];

    if (
        (
            child.type !== 'PunctuationNode' &&
            child.type !== 'SymbolNode'
        ) ||
        nlcstToString(child) !== '.'
    ) {
        return;
    }

    currentIndex = end = start = index;

    /**
     * Find preceding word/punctuation.
     * Stop before slashes, break after `www`.
     */

    while (siblings[start - 1]) {
        type = siblings[start - 1].type;

        if (
            type !== 'WordNode' &&
            type !== 'PunctuationNode' &&
            type !== 'SymbolNode'
        ) {
            break;
        }

        if (
            (
                type === 'PunctuationNode' ||
                type === 'SymbolNode'
            ) &&
            EXPRESSION_SLASH.test(nlcstToString(siblings[start - 1]))
        ) {
            break;
        }

        start--;

        nodes.unshift(siblings[start]);

        if (
            type === 'WordNode' &&
            nlcstToString(siblings[start]) === 'www'
        ) {
            break;
        }
    }

    /**
     * Find following word/punctuation.
     */

    while (siblings[end + 1]) {
        type = siblings[end + 1].type;

        if (
            type !== 'WordNode' &&
            type !== 'PunctuationNode' &&
            type !== 'SymbolNode'
        ) {
            break;
        }

        end++;

        nodes.push(siblings[end]);
    }

    /**
     * This full stop doesnt look like a link:
     * it's either not followed, or not preceded,
     * by words or punctuation.
     */

    if (currentIndex === start || currentIndex === end) {
        return;
    }

    /**
     * 1-3 slashes.
     */

    if (
        start > 0 &&
        siblings[start - 1].type === 'PunctuationNode' &&
        EXPRESSION_SLASH.test(nlcstToString(siblings[start - 1]))
    ) {
        start--;

        nodes.unshift(siblings[start]);
    }

    /**
     * URL protocol and colon.
     */

    if (
        start > 2 &&
        siblings[start - 1].type === 'PunctuationNode' &&
        nlcstToString(siblings[start - 1]) === ':' &&
        siblings[start - 2].type === 'WordNode'
    ) {
        nodes.unshift(siblings[start - 2], siblings[start - 1]);
        start -= 2;
    }

    /**
     * Remove the last node if it's punctuation,
     * unless it's `/` or `)`.
     */

    value = nlcstToString(siblings[end]);

    if (
        siblings[end].type === 'PunctuationNode' &&
        value !== '/' && value !== ')'
    ) {
        end--;

        nodes.pop();
    }

    /**
     * Remove the nodes and insert a LinkNode.
     */

    siblings.splice(start, end - start + 1, {
        'type': LINK_NODE,
        'value': nlcstToString({
            'children': nodes
        })
    });

    index++;

    /**
     * Ignore the following full-stop: it's not part
     * of a link.
     */

    if (value === '.') {
        index++;
    }
}

var linkPlugin;

function attach(parser) {
    if (!parser || !parser.parse) {
        throw new Error(
            '`parser` is not a valid parser for ' +
            '`attach(parser)`. Make sure something ' +
            'like `parse-latin` is passed.'
        );
    }

    /**
     * Make sure to not re-attach the plugin.
     */

    if (!linkPlugin) {
        linkPlugin = parser.constructor.plugin(mergeLinks);
    }

    parser.useFirst('tokenizeSentence', linkPlugin);
}

/**
 * Expose `attach`.
 */

module.exports = attach;


benchmark.js
'use strict';

/**
 * Dependencies.
 */

var modifier,
    ParseEnglish;

modifier = require('./');
ParseEnglish = require('parse-english');

/**
 * `ParseEnglish`.
 */

var parseEnglish,
    linkParseEnglish;

parseEnglish = new ParseEnglish();
linkParseEnglish = new ParseEnglish();

modifier(linkParseEnglish);

/**
 * Fixtures.
 *
 * Source:
 *   http://www.gutenberg.org/cache/epub/11024/pg11024.html
 */

var paragraph,
    linkParagraph;

/**
 * A paragraph, 5 sentences, filled with 10 links.
 */

linkParagraph = 'Thou art a churlish knight to so ' +
    'affront a http://example.com he could not sit ' +
    'upon http://userid:password@example.com/ his ' +
    'horse any longer . ' +

    'For methinks http://j.mp something hath befallen ' +
    'my lord and that he then, foo.com/blah_blah/ ' +
    'after a while, he cried out in great voice. ' +

    'For that 142.42.1.1:8080/ light in the sky lieth ' +
    'in the south a.b-c.de then Queen Helen fell down ' +
    'in a swoon, and lay. ' +

    'Touch me not, مثال.إختبار for I am not ' +
    'mortal, but Fay so the Lady of the Lake vanished ' +
    'away, उदाहरण.परीक्षा ' +
    'everything behind. ' +

    'Where she had http://⌘.ws stood was clear, and ' +
    'she was http://✪df.ws/123 gone since Sir Kay ' +
    'does not choose to assume my quarrel.';

/**
 * A paragraph, 5 sentences, without links.
 */

paragraph = 'Thou art a churlish knight to so affront ' +
    'a lady he could not sit upon his horse any ' +
    'longer. ' +

    'For methinks something hath befallen my lord ' +
    'and that he then, after a while, he cried out ' +
    'in great voice. ' +

    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and ' +
    'lay. ' +

    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, ' +
    'everything behind. ' +

    'Where she had stood was clear, and she was ' +
    'gone since Sir Kay does not choose to assume my ' +
    'quarrel.';
/**
 * Benchmarks.
 */

suite('parse w/ modifier', function () {
    bench('A paragraph (5 sentences, 100 words, 10 links)',
        function () {
            linkParseEnglish.parse(linkParagraph);
        }
    );

    bench('A paragraph (5 sentences, 100 words, no links)',
        function () {
            linkParseEnglish.parse(paragraph);
        }
    );
});

suite('parse w/o modifier', function () {
    bench('A paragraph (5 sentences, 100 words, 10 links)',
        function () {
            parseEnglish.parse(linkParagraph);
        }
    );

    bench('A paragraph (5 sentences, 100 words, no links)',
        function () {
            parseEnglish.parse(paragraph);
        }
    );
});
